helpers/yarn/lib/peer-dependency-checker.js

```diff
@@ -0,0 +1,139 @@
+/* PEER DEPENDENCY CHECKER
+ *
+ * Inputs:
+ *  - directory containing a package.json and a yarn.lock
+ *  - dependency name
+ *  - new dependency version
+ *  - requirements for this dependency
+ *
+ * Outputs:
+ *  - successful completion, or an error if there are peer dependency warnings
+ */
+const path = require("path");
+const { Add } = require("@dependabot/yarn-lib/lib/cli/commands/add");
+const Config = require("@dependabot/yarn-lib/lib/config").default;
+const { EventReporter } = require("@dependabot/yarn-lib/lib/reporters");
+const Lockfile = require("@dependabot/yarn-lib/lib/lockfile").default;
+
+class LightweightAdd extends Add {
+  async bailout(patterns, workspaceLayout) {
+    await this.linker.resolvePeerModules();
+    return true;
+  }
+}
+
+class DependabotReporter extends EventReporter {
+  warn(msg) {
+    if (msg.includes("incorrect peer dependency")) {
+      throw new Error(msg);
+    }
+  }
+}
+
+function devRequirement(requirements) {
+  const groups = requirements.groups;
+  return (
+    groups.indexOf("devDependencies") > -1 &&
+    groups.indexOf("dependencies") == -1
+  );
+}
+
+function optionalRequirement(requirements) {
+  const groups = requirements.groups;
+  return (
+    groups.indexOf("optionalDependencies") > -1 &&
+    groups.indexOf("dependencies") == -1
+  );
+}
+
+function installArgsWithVersion(depName, desiredVersion, requirements) {
+  const source =
+    "source" in requirements
+      ? requirements.source
+      : (requirements.find(req => req.source) || {}).source;
+
+  if (source && source.type === "git") {
+    return [`${depName}@${source.url}#${desiredVersion}`];
+  } else {
+    return [`${depName}@${desiredVersion}`];
+  }
+}
+
+async function checkPeerDependencies(
+  directory,
+  depName,
+  desiredVersion,
+  requirements,
+  topLevelDependencies
+) {
+  for (let req of requirements) {
```

Comment:
@greysteil what is going on here? We don't seem to do this for npm, just get the first requirement with a `source` when figuring out git version.

Do we want to do this for npm? So we get the directory based on the requirement file (for the installer): `directory = path.join(directory, path.dirname(requirement.file))`

Reply:
ah wait that's not even what we do for yarn, is this to deal with workspaces that have a root lockfile?

Reply:
Yeah, exactly that. I'm not sure it's the best way to deal with it - ages since I wrote it :-/